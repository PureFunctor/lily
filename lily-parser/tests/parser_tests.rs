use lily_parser::parse_top_level;

#[test]
pub fn top_level_0() {
    let source = r"
infixl 4 add as +

example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_1() {
    let source = r"
example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_2() {
    let source = r"
infixl 1 add as +
infixl 2 add as *

example = a + b * (c + d) + e
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_3() {
    let source = r"
infixl 1 add as +

example = if if a then b else c then if d then e else f else if g then h else i
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_4() {
    let source = r"
example = f a b
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_5() {
    let source = r"
example = f if a then b else c d

example = f (if a then b else c) d
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}
