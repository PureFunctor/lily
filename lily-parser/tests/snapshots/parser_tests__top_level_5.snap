---
source: lily-parser/tests/parser_tests.rs
expression: parse_top_level(source)
---
Ok(
    Module {
        declarations: [
            Declaration {
                begin: 1,
                end: 33,
                kind: ValueDeclaration(
                    "example",
                    [],
                    Expression {
                        begin: 11,
                        end: 33,
                        kind: Application(
                            Expression {
                                begin: 11,
                                end: 12,
                                kind: Variable(
                                    "f",
                                ),
                            },
                            [
                                Expression {
                                    begin: 13,
                                    end: 33,
                                    kind: IfThenElse(
                                        Expression {
                                            begin: 16,
                                            end: 17,
                                            kind: Variable(
                                                "a",
                                            ),
                                        },
                                        Expression {
                                            begin: 23,
                                            end: 24,
                                            kind: Variable(
                                                "b",
                                            ),
                                        },
                                        Expression {
                                            begin: 30,
                                            end: 33,
                                            kind: Application(
                                                Expression {
                                                    begin: 30,
                                                    end: 31,
                                                    kind: Variable(
                                                        "c",
                                                    ),
                                                },
                                                [
                                                    Expression {
                                                        begin: 32,
                                                        end: 33,
                                                        kind: Variable(
                                                            "d",
                                                        ),
                                                    },
                                                ],
                                            ),
                                        },
                                    ),
                                },
                            ],
                        ),
                    },
                ),
            },
            Declaration {
                begin: 35,
                end: 69,
                kind: ValueDeclaration(
                    "example",
                    [],
                    Expression {
                        begin: 45,
                        end: 69,
                        kind: Application(
                            Expression {
                                begin: 45,
                                end: 46,
                                kind: Variable(
                                    "f",
                                ),
                            },
                            [
                                Expression {
                                    begin: 47,
                                    end: 67,
                                    kind: Parenthesized(
                                        Expression {
                                            begin: 48,
                                            end: 66,
                                            kind: IfThenElse(
                                                Expression {
                                                    begin: 51,
                                                    end: 52,
                                                    kind: Variable(
                                                        "a",
                                                    ),
                                                },
                                                Expression {
                                                    begin: 58,
                                                    end: 59,
                                                    kind: Variable(
                                                        "b",
                                                    ),
                                                },
                                                Expression {
                                                    begin: 65,
                                                    end: 66,
                                                    kind: Variable(
                                                        "c",
                                                    ),
                                                },
                                            ),
                                        },
                                    ),
                                },
                                Expression {
                                    begin: 68,
                                    end: 69,
                                    kind: Variable(
                                        "d",
                                    ),
                                },
                            ],
                        ),
                    },
                ),
            },
        ],
    },
)
